daos-stack/daos
# DAOS-9862 rdb: Introduce local TXs

This patch introduces special "local" RDB TXs for testing and future
catastrophic recovery purposes. Local TXs can be executed with open but
not yet started DBs, represented by rdb_storage objects. These TXs are
query only and see the latest local DB contents that may include
uncommitted updates. In tests, we can therefore verify DB contents on
any replica, who may or may not be a leader.

Here is a summary of the changes:

  - [Core] Add rdb_storage that represents an open but not yet started
    DB.
      - Split "start" into "open" then "start", and "stop" into "stop"
	then "close".

  - Fix a load-snapshot leak on an error path in rdb_raft_load_lc.

  - Add a missing ABT_cond_broadcast(db->d_events_cv) call to an error
    path in rdb_raft_start.

  - Remove some useless ABT_thread_join calls.

  - [Core] Add rdb_tx_begin_local.
      - Use a rdb_tx flag to skip the Raft logic.
      - This requires an rdb_storage object because we may have some
	concurrency problem if we were to execute a local TX that yields
        while there is a concurrent incoming AE request.

  - [Core] Fix rdbt to verify DB contents using local TXs instead of
    leadership transfers, which are not supported.
      - Disable one test case because it requires a leadership transfer.

Signed-off-by: Li Wei <wei.g.li@intel.com>
